File: examples/Iris/iris.cc
#include <iostream>
#include "model.h"
#include "loss.h"
#include "optimizer.h"

#include "iris_dataset.h"

using namespace snnf;
using namespace snnf::layer;
using namespace snnf::dataset;

const int num_epochs = 20;
const float train_pct = 0.8f;
const float learning_rate = 0.01f;

int main() {
    IrisDataset<float> dataset;
    dataset.load_data("data/iris.txt");

    // Load data, set batch size to 16, num_worker to 1, and shuffle.
    DataLoader<float> train_loader(dataset, 16, true, 1);

    Model<float> model;
    model.add_layer(std::make_shared<LinearLayer<float>>(4, 16));
    model.add_layer(std::make_shared<ReLULayer<float>>());
    model.add_layer(std::make_shared<LinearLayer<float>>(16, 3));
    model.add_layer(std::make_shared<SigmoidLayer<float>>());

    SGD<float> optimizer(learning_rate);
    optimizer.add_parameters(model.get_parameters());

    MSELoss<float> loss_fn;

    for (size_t epoch = 0; epoch < num_epochs; ++epoch) {
        train_loader.reset();
        float total_loss = 0.0f;
        size_t batch_count = 0;

        while (train_loader.has_next()) {
            auto batch = train_loader.next_batch();
            auto& input = batch.first;
            auto& target = batch.second;

            optimizer.zero_grad();

            Tensor<float> output = model.forward(input);
            float loss = loss_fn.forward(output, target);

            total_loss += loss;
            batch_count++;

            Tensor<float> loss_grad = loss_fn.backward();
            model.backward(loss_grad);

            optimizer.step();

        }

        std::cout << "Epoch " << epoch + 1 << " of " << num_epochs
                  << ", Loss: " << total_loss / batch_count << std::endl;
    }

    return 0;
}

File: examples/Iris/python/iris.py
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from torch.utils.data import DataLoader, TensorDataset

iris = load_iris()
data = iris.data
labels = iris.target
num_epochs = 100

scaler = StandardScaler()
data = scaler.fit_transform(data)

# 80% of training data.
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)

X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train, dtype=torch.long)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.long)

train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(4, 16)
        self.fc2 = nn.Linear(16, 8)
        self.fc3 = nn.Linear(8, 3)
    
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

model = Net()
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)
for epoch in range(num_epochs):
    model.train()
    total_loss = 0
    for X_batch, y_batch in train_loader:
        optimizer.zero_grad()
        outputs = model(X_batch)
        loss = criterion(outputs, y_batch)
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    if (epoch + 1) % 10 == 0:
        print(f"Epoch [{epoch + 1}/{epochs}], Loss: {total_loss:.4f}")

model.eval()
with torch.no_grad():
    y_pred = model(X_test_tensor)
    predicted = torch.argmax(y_pred, dim=1)
    accuracy = (predicted == y_test_tensor).sum().item() / y_test_tensor.size(0)
    print(f"Test Accuracy: {accuracy * 100:.2f}%")